src/txrequest.cpp

```diff
@@ -0,0 +1,593 @@
+struct Announcement {
+    /** Txid or wtxid that was announced. */
+    const uint256 m_txhash;
+    /** For CANDIDATE_{DELAYED,BEST,READY} the reqtime; for REQUESTED the expiry. */
+    std::chrono::microseconds m_time;
+    /** What peer the request was from. */
+    const NodeId m_peer;
+    /** What sequence number this announcement has. */
+    const SequenceNumber m_sequence : 59;
+    /** Whether the request is preferred. */
+    const bool m_preferred : 1;
+    /** Whether this is a wtxid request. */
+    const bool m_is_wtxid : 1;
+
+    /** What state this announcement is in. */
+    State m_state : 3;
+
+    /** Whether this announcement is selected. There can be at most 1 selected peer per txhash. */
+    bool IsSelected() const
+    {
+        return m_state == State::CANDIDATE_BEST || m_state == State::REQUESTED;
+    }
+
+    /** Whether this announcement is waiting for a certain time to pass. */
+    bool IsWaiting() const
+    {
+        return m_state == State::REQUESTED || m_state == State::CANDIDATE_DELAYED;
+    }
+
+    /** Whether this announcement can feasibly be selected if the current IsSelected() one disappears. */
+    bool IsSelectable() const
+    {
+        return m_state == State::CANDIDATE_READY || m_state == State::CANDIDATE_BEST;
+    }
+
+    /** Construct a new announcement from scratch, initially in CANDIDATE_DELAYED state. */
+    Announcement(const GenTxid& gtxid, NodeId peer, bool preferred, std::chrono::microseconds reqtime,
+        SequenceNumber sequence) :
+        m_txhash(gtxid.GetHash()), m_time(reqtime), m_peer(peer), m_sequence(sequence), m_preferred(preferred),
+        m_is_wtxid(gtxid.IsWtxid()), m_state(State::CANDIDATE_DELAYED) {}
+};
+
+//! Type alias for priorities.
+using Priority = uint64_t;
+
+/** A functor with embedded salt that computes priority of an announcement.
+ *
+ * Higher priorities are selected first.
+ */
+class PriorityComputer {
+    const uint64_t m_k0, m_k1;
+public:
+    explicit PriorityComputer(bool deterministic) :
+        m_k0{deterministic ? 0 : GetRand(0xFFFFFFFFFFFFFFFF)},
+        m_k1{deterministic ? 0 : GetRand(0xFFFFFFFFFFFFFFFF)} {}
+
+    Priority operator()(const uint256& txhash, NodeId peer, bool preferred) const
+    {
+        uint64_t low_bits = CSipHasher(m_k0, m_k1).Write(txhash.begin(), txhash.size()).Write(peer).Finalize() >> 1;
+        return low_bits | uint64_t{preferred} << 63;
+    }
+
+    Priority operator()(const Announcement& ann) const
+    {
+        return operator()(ann.m_txhash, ann.m_peer, ann.m_preferred);
+    }
+};
+
+// Definitions for the 3 indexes used in the main data structure.
+//
+// Each index has a By* type to identify it, a By*View data type to represent the view of announcement it is sorted
+// by, and an By*ViewExtractor type to convert an announcement into the By*View type.
+// See https://www.boost.org/doc/libs/1_54_0/libs/multi_index/doc/reference/key_extraction.html#key_extractors
```

Comment:
1950db7598a426232aeb5b6e1114a1f7e1ab35a1, nit:
```suggestion
// See https://www.boost.org/doc/libs/1_58_0/libs/multi_index/doc/reference/key_extraction.html#key_extractors
```

Reply:
Done.